ucm2/sof-soundwire/rt722-0.conf

```diff
@@ -0,0 +1,18 @@
+# Use case Configuration for sof-soundwire card
+
+SectionDevice."Speaker" {
+        Comment "Speaker"
+
+        EnableSequence [
+                cset "name='Speaker Switch' on"
+        ]
+
+        DisableSequence [
+                cset "name='Speaker Switch' off"
+        ]
+
+        Value {
+              PlaybackPriority 100
+              PlaybackPCM "hw:${CardId},2"
+        }
+}
```

Comment:
the file naming is confusing, why not use the function name, i.e.

rt722-amp.conf
rt722-jack.conf
rt722-mic.conf

Reply:
Because the speaker structure defined in HiFi.conf is like below

If.spkdev {
	Condition {
		Type String
		Empty "${var:SpeakerCodec1}"
	}
	False.Include.spkdev.File "/sof-soundwire/${var:SpeakerCodec1}-${var:SpeakerAmps1}.conf"
}

micdev and hsdev also has its structure defined in HiFi.conf, so I follow these just like what rt712 did.

Reply:
rt722-0.conf means nothing to me...

Reply:
@bardliao @jack-cy-yu I wonder if the amixer 'components' string is correct now that I think of it, could it be that we report zero amps?

Can you share the result of amixer -Dhw:L info ?

Reply:
@plbossart 
Card hw:0 'sofsoundwire'/'Intel Soundwire SOF'
  Mixer name	: 'Intel Meteor Lake HDMI'
  Components	: 'HDA:8086281d,80860101,00100000 cfg-spk:2 cfg-amp:0 spk:rt722 hs:rt722-sdca mic:rt722'
  Controls      : 21
  Simple ctrls  : 16

Reply:
The driver serves one chip or two chips (rt722 / rt722-sdca) ?

Reply:
> The driver serves one chip or two chips (rt722 / rt722-sdca) ?

One chip, but currently, machine driver will add "-sdca" suffix for headset.

Reply:
@bardliao this configuration makes no sense "cfg-spk:2 cfg-amp:0". We're missing something here, the amps cannot be zero. and I thought we stopped reporting cfg-spk?

Reply:
I think that we should end with one `rt722.conf` file and use a different identifier in components (like `cfg-codec:rt722`) to not follow the scheme with separate chips. All other components without use should be removed.

Reply:
@perexg "cfg-codec:rt722" may only work if this is the only chip in the system. I vaguely remember the possibility of having an additional discrete amplifier in addition to RT722 for e.g. tweeters.

With SDCA the 'functions' may be part of a single chip or a set of discrete ones. There's no general rule on what the partition might be.

Reply:
> @bardliao this configuration makes no sense "cfg-spk:2 cfg-amp:0". We're missing something here, the amps cannot be zero. and I thought we stopped reporting cfg-spk?

The output is from @jack-cy-yu. He may not use the latest sof-dev kernel. The chg-amp:0 issue was fixed by https://github.com/thesofproject/linux/pull/4909

Reply:
> @perexg "cfg-codec:rt722" may only work if this is the only chip in the system. I vaguely remember the possibility of having an additional discrete amplifier in addition to RT722 for e.g. tweeters.
> 
> With SDCA the 'functions' may be part of a single chip or a set of discrete ones. There's no general rule on what the partition might be.

We can extend rt722.conf (add include) or the sof-soundwire.conf file when we know this hardware. From my view, trying to separate functions for one driver seems a bit more difficult than maintain whole configuration in one file.

We can do something like (assuming that all other codec indentifications will be removed from components string, keeping only `codec:rt722` there:

```
diff --git a/ucm2/sof-soundwire/HiFi.conf b/ucm2/sof-soundwire/HiFi.conf
index ffea077..8777ee6 100644
--- a/ucm2/sof-soundwire/HiFi.conf
+++ b/ucm2/sof-soundwire/HiFi.conf
@@ -8,6 +8,14 @@ SectionVerb {
        Value.TQ "HiFi"
 }
 
+If.codec {
+       Condition {
+               Type String
+               Empty "${var:Codec1}"
+       }
+       False.Include.codec.File "/sof-soundwire/${var:Codec1}.conf"
+}
+
 If.spkdev {
        Condition {
                Type String
diff --git a/ucm2/sof-soundwire/sof-soundwire.conf b/ucm2/sof-soundwire/sof-soundwire.conf
index 35a262c..1ed888c 100644
--- a/ucm2/sof-soundwire/sof-soundwire.conf
+++ b/ucm2/sof-soundwire/sof-soundwire.conf
@@ -18,6 +18,10 @@ Define {
 }
 
 DefineRegex {
+       Codec {
+               Regex " codec:([a-z0-9]+(-sdca)?)"
+               String "${CardComponents}"
+       }
        SpeakerCodec {
                Regex " spk:([a-z0-9]+(-sdca)?)"
                String "${CardComponents}"
@@ -44,6 +48,15 @@ DefineRegex {
        }
 }
 
+If.codec_init {
+       Condition {
+               Type RegexMatch
+               Regex "(rt722)"
+               String "${var:Codec1}"
+       }
+       True.Include.codec_init.File "/codecs/${var:Codec1}/init.conf"
+}
+
 If.hs_init {
        Condition {
```

Reply:
I've followed the suggestion of @perexg  and added the patch like attachment, it works on my MTL platform, but I declare the variable "Codec1" from 

 DefineRegex {
+	Codec {
+		Regex " spk:([a-z0-9]+(-sdca)?)"
+		String "${CardComponents}"
+	}
 ...
 }

we need "Regex " **codec**:([a-z0-9]+(-sdca)?)" " to replace "Regex " **spk**:([a-z0-9]+(-sdca)?)"".  @plbossart @bardliao Could you help to add a "codec:rt722" in components string? 

[0001-ucm2-soundwire-add-rt722-SDCA-device.zip](https://github.com/alsa-project/alsa-ucm-conf/files/15091751/0001-ucm2-soundwire-add-rt722-SDCA-device.zip)

Reply:
> we need "Regex " **codec**:([a-z0-9]+(-sdca)?)" " to replace "Regex " **spk**:([a-z0-9]+(-sdca)?)"". @plbossart @bardliao Could you help to add a "codec:rt722" in components string?
> 

Yes, I will submit a machine driver change for it.

Reply:
I really didn't get how we will support the case with two chips for the main/companion amplifier topology. It's not an hypothetical design, it's what OEMs want to scale their platforms between mid/high tiers.

Reply:
The driver should set a hint for UCM in the components string and UCM configuration should be extended for both cases. There are several possibilities to describe this situation, but I guess, we should do it when we have a working driver.

Reply:
I am fine with pushing the main/companion amplifier work to later, but the term 'codec' means absolutely nothing in the case of RT722. It's a multi-function device, so we don't know if we are referring to the jack, amp or mic function, or all of the above.

Reply:
Judging from the rt722-sdca.c source file (datasheet is not probably publicly available), the comments in `rt722_sdca_controls` table are straight forward about the expected usage. I don't comment the ugly control names transferring any abstraction to upper layers. We lost this abstraction from 0day for ASoC drivers.